Nested set and foreign keys

jraddaoui opened the pull request:
Refs #13222, #13237, #13313 and #10306.

See commit messages for more info.

jraddaoui committed:
Fix culture columns in table maps, refs #13222

Return their size to 16 after the changes from b8206cf.

jraddaoui committed:
Remove nested set from some models, refs #13237

- Remove parent_id, lft and rgt columns from:
  - QubitFunctionObject.
  - QubitPhysicalObject.
- Remove lft and rgt columns from:
  - QubitActor (and other models extending this one).
  - QubitAclGroup.
  - QubitTaxonomy.

With this changes, the only models with a full nested set implementation
are QubitInformationObject, QubitTerm and QubitMenu as they require a
method to order siblings. While QubitActor (and related models),
QubitAclGroup and QubitTaxonomy only have a parent_id column, used for
the ACL system alongside CTE in the ancestors queries. Additionally,
the QubitDigitalObject model has only a parent_id column, in this case
to relate the derivatives with their master.

- Modify schema.
- Build SQL.
- Add migration.
- Bump version.
- Add addGetAncestorsAndSelfForAcl to QubitObjectBuilder:
  - Based on the self-relationship using parent_id.
  - Needed for the ACL system to get ancestors in those models that
    don't have a full nested set implementation.
  - Also used for models with nested set as it's more efficient and
    simplifies the code.
  - Uses a join with a CTE and level order to replicate the ordered
    results from `ancestors->andSelf()->orderBy('lft')`.
- Build models.
- Remove calls to non existing methods.
- Remove not needed uses and references of lft and rgt columns.
- Replace QubitActor by QubitMenu in the build nested set task.

jraddaoui committed:
Modify MySQL's sql_mode in migrations, refs #13313

Read code comment.

jraddaoui commented:
Hi @sbreker and @djjuhasz, while the PR is still in progress (I'll address #13239 in here too) the current commits relate to other issues and you could start reviewing if you have some time. Thanks!

djjuhasz commented on config/schema.yml:
```diff
@@ -287,9 +287,9 @@ propel:
-    type_id: { type: integer, foreignTable: term, foreignReference: id }
+    subject_id: { type: integer, required: true, foreignTable: object, foreignReference: id, onDelete: cascade }
+    object_id: { type: integer, required: true, foreignTable: object, foreignReference: id, onDelete: cascade }
+    type_id: { type: integer, foreignTable: term, foreignReference: id, onDelete: cascade }
```
This one worries me a bit. In principle I think it's true that a relation without at `type_id` doesn't mean much, my concern is that a user deletes a term without realizing it will also delete all related relation rows.  It's a tough choice but I lean slightly towards "onDelete: setnull" for type_id to avoid unintended relation deletes.

jraddaoui replied:
I kind of agree with @djjuhasz but I'd like to hear your opinion in this one, @fiver-watson.

fiver-watson replied:
Not sure I'm totally understanding this... What would be the advantage of leaving a relation row that points to a NULL, @djjuhasz  - can you think of real-world examples where someone would want to delete a term but still have references to the missing term in the related entities? And, would the user have any option for deleting that relation once the term is gone? Thanks1

sbreker replied:
Current functionality appears to be 'ON DELETE RESTRICT' which will prevent deletion of terms for which relations exist.

Dan raises a good point - if relation.type_id is NULL, then the relation will be ignored in any queries that filter on types of relations.

djjuhasz replied:
@fiver-watson I can't think of any use case where having a relation row with no type_id is desirable.  

My concern is that somebody deleting a term (especially via direct MySQL access rather than the UI) that is used for a relation.type_id, without realizing that it's going to delete every relationship that uses that term.  E.g. a user naively deletes the "controls" term, and wipes out all "controls" relationships.  At least if the relation.type_id is set to null, the user can recover by adding a new type_id to the row. After deleting the term, the user could delete all relation rows where relation.type_id = NULL if that's the objective.

It's a tough call though, allowing relation.type_id to be null does have the potential to hide those relation rows from the UI and in effect make them hidden data.

fiver-watson replied:
I've talked about this a bit more with @jraddaoui and we are both still kind of leaning towards ON DELETE CASCADE. In the UI, we at least give users a warning page before a deletion (though with terms we don't give a count of affected relations). In the DB, we **always** recommend making a backup before directly manipulating the data using SQL - i'm not sure that leaving empty rows to account for potentially undoing user error is a good tradeoff in terms of overall database consistency. 

We could potentially consider rewording the warning screen that appears when a term is deleted, to add a clearer warning? Even without a count, that might be useful. 

Ultimately, i think this is a decision that I trust you all to come to consensus around, as it feels more about project maintenance than end user experience and usage. I think your argument is an interesting one David, but what would it mean if we applied it consistently throughout AtoM?

djjuhasz replied:
Okay, I'm fine with ON DELETE CASCADE - I agree that it's better for database consistency and I'm probably being over-cautious in worrying about unanticipated deletes.

djjuhasz replied:
On the topic of database consistency, I wonder if it's possible to do an ON DELETE CASCADE of the `object` row when a related object is deleted?  I think it would require putting a foreign key constraint on e.g. `information_object.id` to `object.id` which may cause a circular dependency, but it would sure be nice to avoid leaving behind useless `object` table rows.

jraddaoui replied:
Hi @djjuhasz, I'm not sure how that would work in the ORM but I think @fiver-watson added some notes about this in the SQL work he did. I'd prefer not to go that way and ask to do the deletes from the object row, otherwise we may want to do the same in other inheritances (QubitRepository and others from QubitActor for example) and it could get messy.

jraddaoui replied:
And thanks all for the discussion.

djjuhasz commented on lib/task/migrate/migrations/arMigration0177.class.php:
```diff
@@ -75,102 +75,27 @@ public function up($configuration)
+        'column' => 'object_id',
+        'refTable' => 'object',
+        'constraint' => 'digital_object_FK_2',
+        'onDelete' => ' ON DELETE CASCADE',
```
I'd prefer to see the leading space in the SQL templates, rather than having it in the 'onDelete' value.  To me it seems too easy to forgot to add the leading space when adding new migrations.

jraddaoui replied:
Great point!

djjuhasz commented on lib/propel/builder/QubitObjectBuilder.php:
```diff
@@ -1924,6 +1929,46 @@ public function hasChildren()
+      )";
+
+      \$criteria = new Criteria;
+      \$criteria->addJoin({$selfFkRefColCons}, 'cte.id', "RIGHT JOIN \$cte");
```
Huh, this is cool - I didn't realize you could use raw SQL in a Propel Criteria join like this. 😮

jraddaoui replied:
Me neither :smile:! It wasn't easy to figure out and I feel like I'm hacking it, but it works :stuck_out_tongue:.

sbreker replied:
This is going to really help the digital object browse page!

jraddaoui replied:
Thanks @sbreker!

We were already using CTE but it was in a subquery so this should be a bit better. Additionally, it removes the extra processing from `andSelf()` and `orderBy()`:

https://github.com/artefactual/atom/blob/qa/2.6.x/lib/QubitQuery.class.php#L137-L170

:rocket:

djjuhasz commented on lib/task/propel/propelBuildNestedSetTask.class.php:
```diff
@@ -61,8 +61,8 @@ public function execute($arguments = array(), $options = array())
-      'actor' => 'QubitActor',
-      'term' => 'QubitTerm'
+      'term' => 'QubitTerm',
+      'menu' => 'QubitMenu'
```
lol, I can't believe QubitMenu was left out of this task for so long! 😆

djjuhasz commented on plugins/qbAclPlugin/lib/QubitAcl.class.php:
```diff
@@ -269,7 +269,7 @@ protected function addRole($role)
       if (!in_array($role->id, $this->_roles))
       {
-        foreach ($role->ancestors->andSelf()->orderBy('lft') as $ancestor)
+        foreach ($role->getAncestorsAndSelfForAcl() as $ancestor)
```
Awesome! 😁

djjuhasz reviewed (commented):
This is looking great so far @jraddaoui!  A few questions/comments but no blockers. 👍

jraddaoui commented:
Thanks for the quick and great feedback @djjuhasz!

sbreker commented:
Great work so far @jraddaoui!

jraddaoui commented:
Hi @djjuhasz and @sbreker,

Please, let me know if you think this is ready for merge. I'll address #13239 in a different PR in the end to allow testing this changes in the meantime.

Thanks!

PS: I'll squash only the last two commits if it's ready and merge with `--ff-only`.

djjuhasz reviewed (approved):
Looks good to me. :)

jraddaoui committed:
Add on delete actions to foreign keys, refs #10306

Add missing on delete actions to:
- actor:
  - parent_id: CASCADE
- digital_object:
  - parent_id: CASCADE
- event:
  - actor_id: SET NULL
- function_object:
  - type_id: SET NULL
  - description_status_id: SET NULL
  - description_detail_id: SET NULL
- information_object:
  - collection_type_id: SET NULL
  - repository: SET NULL
  - parent_id: CASCADE
- note:
  - user_id: SET NULL
- premis_object
  - information_object_id: CASCADE
- relation:
  - subject_id: CASCADE
  - object_id: CASCADE
  - type_id: CASCADE
- taxonomy:
  - parent_id: CASCADE
- term:
  - parent_id: CASCADE

Changes:
- Modify schema.
- Build SQL.
- Build models.
- Add migration:
  - Normalize update indexes and foreign keys functions in QubitMigrate.
  - Use them from migrations: 177, 181 and 182.
  - Catch exception updating foreign key to reformat message with
    more information about the failed operation.
- Bump version.
